### PR TITLE
ENT-2650 - adding logic in CouponRedeemView to handle 100% off Enterprise Coupons

### PR DESCRIPTION
**Pull request by chavesj**
[Ticket](https://openedx.atlassian.net/browse/ENT-2650)

### CODE REVIEW

**Comment by chavesj** [comment]
Redirect takes user to the courseware when 100% coupon is applied. I have run into issues with local session info because ecommerce uses localhost in some places and edx.devstack in others. 

It seems the purchase was not fully processed since when the user logs back in the prompt to upgrade to a verified course is still displayed - but when that flow is initiated as error is displayed indicating the user already purchased that product.

**Comment by chavesj** [comment]
**Updated Status on this PR**
_post_to_enrollment_api is failing here: https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/fulfillment/modules.py#L423

The stack trace in the Ecommerce logs aren't very helpful since the LMS API Client is making the call. Here are the relevant portions of each stack trace:
**Ecommerce:**
`
edx.devstack.ecommerce    | 2020-03-10 17:18:24,977 ERROR 14 [ecommerce.extensions.fulfillment.modules] /edx/app/ecommerce/ecommerce/ecommerce/extensions/fulfillment/modules.py:452 - Fulfillment of line [31] on order [EDX-100055] failed with status code [400]: An error occurred while creating the new course enrollment for user 'jefftest' in course 'course-v1:edX+DemoX+Demo_Course'
`
**LMS:**
`
edx.devstack.lms          | [10/Mar/2020 17:18:15] "POST /enterprise/api/v1/enterprise-learner/ HTTP/1.1" 201 100
edx.devstack.lms          | 2020-03-10 17:18:24,726 INFO 654 [tracking] [user None] logger.py:49 - {"ip": "172.18.0.1", "referer": "", "username": "", "event_type": "/api/enrollment/v1/enrollment", "time": "2020-03-10T17:18:24.725996+00:00", "page": null, "agent": "python-requests/2.23.0", "accept_language": "", "event_source": "server", "host": "edx.devstack.lms:18000", "event": "{\"POST\": {}, \"GET\": {}}", "context": {"course_id": "", "path": "/api/enrollment/v1/enrollment", "org_id": "", "user_id": null}}
edx.devstack.lms          | 2020-03-10 17:18:24,731 INFO 654 [openedx.core.lib.log_utils] [user None] log_utils.py:37 - ApiKeyHeaderPermission used: ip="172.18.0.1", path="/api/enrollment/v1/enrollment"
edx.devstack.lms          | 2020-03-10 17:18:24,737 INFO 654 [openedx.core.lib.log_utils] [user None] log_utils.py:37 - ApiKeyHeaderPermission used: ip="172.18.0.1", path="/api/enrollment/v1/enrollment"
edx.devstack.lms          | 2020-03-10 17:18:24,738 INFO 654 [openedx.core.lib.log_utils] [user None] log_utils.py:37 - ApiKeyHeaderPermission used: ip="172.18.0.1", path="/api/enrollment/v1/enrollment"
edx.devstack.lms          | 2020-03-10 17:18:24,786 INFO 654 [tracking] [user None] logger.py:49 - {"ip": "127.0.0.1", "referer": "", "username": "", "event_type": "/enterprise/api/v1/enterprise-course-enrollment/", "time": "2020-03-10T17:18:24.786787+00:00", "page": null, "agent": "python-requests/2.22.0 edx-rest-api-client/3.0.2 172.18.0.13", "accept_language": "", "event_source": "server", "host": "localhost:18000", "event": "{\"POST\": {}, \"GET\": {}}", "context": {"course_id": "", "path": "/enterprise/api/v1/enterprise-course-enrollment/", "org_id": "", "user_id": null}}
edx.devstack.lms          | [10/Mar/2020 17:18:24] "POST /enterprise/api/v1/enterprise-course-enrollment/ HTTP/1.1" 403 63
edx.devstack.lms          | 2020-03-10 17:18:24,898 ERROR 654 [edx.enterprise_helpers] [user None] api.py:171 - An error occured while posting EnterpriseCourseEnrollment for user jefftest31 and course run course-v1:edX+DemoX+Demo_Course (consent_granted value: None)
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 161, in post_enterprise_course_enrollment
edx.devstack.lms          |     endpoint.post(data=data)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 167, in post
edx.devstack.lms          |     resp = self._request("POST", data=data, files=files, params=kwargs)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 101, in _request
edx.devstack.lms          |     raise exception_class("Client Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)
edx.devstack.lms          | slumber.exceptions.HttpClientError: Client Error 403: http://localhost:18000/enterprise/api/v1/enterprise-course-enrollment/
edx.devstack.lms          | 2020-03-10 17:18:24,901 ERROR 654 [openedx.core.djangoapps.enrollments.views] [user None] views.py:751 - An unexpected error occurred while creating the new EnterpriseCourseEnrollment for user [jefftest31] in course run [course-v1:edX+DemoX+Demo_Course]
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 161, in post_enterprise_course_enrollment
edx.devstack.lms          |     endpoint.post(data=data)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 167, in post
edx.devstack.lms          |     resp = self._request("POST", data=data, files=files, params=kwargs)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 101, in _request
edx.devstack.lms          |     raise exception_class("Client Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)
edx.devstack.lms          | slumber.exceptions.HttpClientError: Client Error 403: http://localhost:18000/enterprise/api/v1/enterprise-course-enrollment/
edx.devstack.lms          |
edx.devstack.lms          | During handling of the above exception, another exception occurred:
edx.devstack.lms          |
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/enrollments/views.py", line 748, in post
edx.devstack.lms          |     enterprise_api_client.post_enterprise_course_enrollment(username, text_type(course_id), None)
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 172, in post_enterprise_course_enrollment
edx.devstack.lms          |     raise EnterpriseApiException(message)
edx.devstack.lms          | openedx.features.enterprise_support.api.EnterpriseApiException: An error occured while posting EnterpriseCourseEnrollment for user jefftest31 and course run course-v1:edX+DemoX+Demo_Course (consent_granted value: None)
edx.devstack.lms          | 2020-03-10 17:18:24,902 ERROR 654 [openedx.core.djangoapps.enrollments.views] [user None] views.py:844 - An error occurred while creating the new course enrollment for user [jefftest31] in course run [course-v1:edX+DemoX+Demo_Course]
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 161, in post_enterprise_course_enrollment
edx.devstack.lms          |     endpoint.post(data=data)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 167, in post
edx.devstack.lms          |     resp = self._request("POST", data=data, files=files, params=kwargs)
edx.devstack.lms          |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/slumber/__init__.py", line 101, in _request
edx.devstack.lms          |     raise exception_class("Client Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)
edx.devstack.lms          | slumber.exceptions.HttpClientError: Client Error 403: http://localhost:18000/enterprise/api/v1/enterprise-course-enrollment/
edx.devstack.lms          |
edx.devstack.lms          | During handling of the above exception, another exception occurred:
edx.devstack.lms          |
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/enrollments/views.py", line 748, in post
edx.devstack.lms          |     enterprise_api_client.post_enterprise_course_enrollment(username, text_type(course_id), None)
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 172, in post_enterprise_course_enrollment
edx.devstack.lms          |     raise EnterpriseApiException(message)
edx.devstack.lms          | openedx.features.enterprise_support.api.EnterpriseApiException: An error occured while posting EnterpriseCourseEnrollment for user jefftest31 and course run course-v1:edX+DemoX+Demo_Course (consent_granted value: None)
edx.devstack.lms          |
edx.devstack.lms          | During handling of the above exception, another exception occurred:
edx.devstack.lms          |
edx.devstack.lms          | Traceback (most recent call last):
edx.devstack.lms          |   File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/enrollments/views.py", line 752, in post
edx.devstack.lms          |     raise CourseEnrollmentError(text_type(error))
edx.devstack.lms          | openedx.core.djangoapps.enrollments.errors.CourseEnrollmentError: An error occured while posting EnterpriseCourseEnrollment for user jefftest31 and course run course-v1:edX+DemoX+Demo_Course (consent_granted value: None)
edx.devstack.lms          | 2020-03-10 17:18:24,939 INFO 654 [openedx.core.lib.log_utils] [user None] log_utils.py:37 - enrollment_change_requested: actual_activation="True", actual_mode="audit", course_id="course-v1:edX+DemoX+Demo_Course", requested_activation="True", requested_mode="verified", user_id="44"
edx.devstack.lms          | [10/Mar/2020 17:18:24] "POST /api/enrollment/v1/enrollment HTTP/1.1" 400 138
`

**Comment by chavesj** [comment]
I initially intended on splitting the [test_enterprise_customer_successful_redemption](https://github.com/edx/ecommerce/blob/chavesj/ENT-2650/ecommerce/coupons/tests/test_views.py#L645) into two tests but I soon realized that the redirect for a non-enterprise 100% coupon is tested in [test_basket_redirect_enrollment_code](https://github.com/edx/ecommerce/blob/chavesj/ENT-2650/ecommerce/coupons/tests/test_views.py#L482). 
My last change updates the test_enterprise_customer_successful_redemption test to expect the courseware URL.

File: ecommerce/coupons/views.py [hunk 1]
```diff
@@ -262,6 +262,10 @@ def get(self, request):  # pylint: disable=too-many-statements
         if basket.total_excl_tax == 0:
             try:
                 order = self.place_free_order(basket)
+                if enterprise_customer:
+                    course_run_id = order.lines.all()[:1].get().product.course.id
```
Comment (bbaker6225) [inline ecommerce/coupons/views.py]: You should be able to do `order.lines.first().product.course.id` to achieve the same thing I believe

Comment (chavesj) [reply ecommerce/coupons/views.py]: I was just doing this the same way it's done [here](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/checkout/views.py#L78)
I'll change in both places

File: ecommerce/coupons/tests/test_views.py [hunk 2]
```diff
@@ -655,10 +655,10 @@ def test_enterprise_customer_successful_redemption(self):
             ENTERPRISE_CUSTOMER
         )
 
-        self.assert_redirects_to_receipt_page(
-            code=code,
-            consent_token=consent_token
-        )
+        response = self.redeem_coupon(code=code, consent_token=consent_token)
```
Comment (adamstankiewicz) [inline ecommerce/coupons/tests/test_views.py:659]: I might update the comment of this test and/or the name of the test to make it more explicit that we're testing the 100% off flow here.

**Review by adamstankiewicz — APPROVED** [review]
Looks good to me. Just one minor suggestion to clarify what `test_enterprise_customer_successful_redemption` is testing (i.e., 100% off code), and incorporating Brandon's syntax feedback.

**Review by brittneyexline — APPROVED** [review]
looks good, though I second Adam and Brandon's feedback. another nit: change the title in both this PR and the commit message (if applicable) to remove WIP and add the jira ticket number (ie ENT-1234).

**Pull request merged**